jpeg.go

```diff
@@ -5,49 +5,156 @@ import (
 	"encoding/binary"
 	"errors"
 	"io"
+	"io/ioutil"
 )
 
 var headerBytes []byte = []byte{0xff, 0xd8}
 
+// scanReader reads until the scan terminating bytes are encountered
+type scanReader struct {
+	reader *io.Reader
+}
+
+func firstMarker(p []byte) int {
+	markerIndex := -1
+	// Look for an 0xff followed by anything other than 0x00 or [0xd0-0xd7] (those are RST markers)
+	for i := 0x01; i <= 0xff; i++ {
+		if i >= 0xd0 && i <= 0xd7 {
+			continue
+		}
+		markerIndex = bytes.Index(p, []byte{0xff, byte(i)})
+		if markerIndex >= 0 {
+			return markerIndex
+		}
+	}
+	return markerIndex
+}
+
+func (sr *scanReader) Read(p []byte) (n int, err error) {
+	buf := new(bytes.Buffer)
+	teedReader := io.TeeReader(*sr.reader, buf)
+	// kick the underlying reader into giving us more than just one byte because the
+	// the naive multireader will just give us one byte from the first reader in its arguments
+	if len(p) > 1 {
+		if n, err = io.ReadAtLeast(teedReader, p, 2); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			} else if errors.Is(err, io.ErrUnexpectedEOF) {
+				// safe to just deal with on the next pass through
+				err = nil
+			} else {
+				err = &MalformedDataError{"unexpected error reading inside scan data", err}
+				return
+			}
+		}
+	} else {
+		// if len(p) is only 1, we can still attempt to read because we'll still
+		// peak the next byte and can act accordingly
+		if n, err = teedReader.Read(p); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			}
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	// attempt to peek one more byte to ensure that the marker doesn't overlap
+	// the bytes returned
+	lastByte := make([]byte, 1)
+	var lastByteN int
+	if lastByteN, err = io.ReadFull(teedReader, lastByte); err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			// an EOF is ok here because we'll catch it normally on the next pass through
+			err = nil
+		} else {
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	var markerIndex int
+	if lastByteN == 1 {
+		markerIndex = firstMarker(append(p[:n], lastByte[0]))
+	} else {
+		markerIndex = firstMarker(p[:n])
+	}
+
+	if markerIndex >= 0 {
+		n = markerIndex
+		if _, err = io.CopyN(ioutil.Discard, buf, int64(markerIndex)); err != nil {
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+
+		// pop "unread" bytes back onto reader
+		*sr.reader = io.MultiReader(buf, *sr.reader)
+		if markerIndex == 0 {
+			err = io.EOF
```

Comment:
What if the read happens to start on a marker?  I think that'd mean `markerIndex == 0`, but there's still data to be read.

Reply:
If the read starts on a marker, then the read should be empty and it should return an EOF. Lemme see if I can diagram a couple examples. Let's say for convenience the marker is `xffxff` and here's the byte pattern sequence in the underlying reader:
`x12x23x34x45x56x67x78x89x90xffxffxfexedxbexef`
Then repeatedly calling `Read` on the `scanReader` should eventually leave you with:
`x12x23x34x45x56x67x78x89x90` followed by an EOF
However, if the underlying reader is:
`xffxffxfexedxbexef`
Then the first `Read` call the `scanReader` should return `0, EOF`

fwiw, this is an attempt to copy as well as I can the semantics of [io.LimitReader](https://golang.org/pkg/io/#LimitedReader)

Reply:
Ahh I see.  I lost track of the context here (that `scanReader` is returned as a segment reader and should read a single segment).  Cool, thanks for the explanation, I appreciate it